src/protocol/boolean/bitwise_gt_constant.rs

```diff
@@ -0,0 +1,262 @@
+use super::into_bits;
+use super::or::or;
+use crate::error::Error;
+use crate::ff::Field;
+use crate::protocol::{context::Context, BitOpStep, RecordId};
+use crate::secret_sharing::Arithmetic as ArithmeticSecretSharing;
+
+/// Compares the `[a]` and `c`, and returns `1` iff `a > c`
+///
+/// Rabbit: Efficient Comparison for Secure Multi-Party Computation
+/// 2.1 Comparison with Bitwise Shared Input – `LTBits` Protocol
+/// Eleftheria Makri, et al.
+/// <https://eprint.iacr.org/2021/119.pdf>
+///
+/// ## Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+pub async fn bitwise_greater_than_constant<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    a: &[S],
+    c: F,
+) -> Result<S, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: ArithmeticSecretSharing<F>,
+{
+    let c_bits = into_bits(c);
+    let first_diff_bit = first_differing_bit(&ctx, record_id, a, &c_bits).await?;
+
+    // Compute the dot-product [a] x `first_diff_bit`. 1 iff a > c.
+    // We can swap `a` with `c` to yield 1 iff a < c. We just need to convert
+    // `c` to `&[c]` using `local_secret_shared_bits`.
+    ctx.narrow(&Step::DotProduct)
+        .sum_of_products(record_id, &first_diff_bit, a)
+        .await
+}
+
+async fn first_differing_bit<F, C, S>(
+    ctx: &C,
+    record_id: RecordId,
+    a: &[S],
+    b: &[F],
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: ArithmeticSecretSharing<F>,
+{
+    let one = ctx.share_of_one();
+
+    // Compute `[a] ^ b`. This step gives us the bits of values where they differ.
+    let mut xored_bits = std::iter::zip(a, b)
+        .map(|(a_bit, b_bit)| {
+            // Local XOR operation `S ^ F`:
+            //    [a] ^ b
+            //  = a_1 - (2 * a_1 * b) + a_2 - (2 * a_2 * b) + a_3 - (2 * a_3 * b) + b
+            //
+            // There's `+ b` at the end, but we don't have Add ops impl for `S + F`.
+            // We need to do a little trick here.
+            //
+            // If `b` = 1, then `+ 3` which is 1 in Fp2.
+            // If `b` = 0, then `+ 0`.
+            //
+            // This is the same as adding a local share `[b] = b == 1 ? one : zero`.
+            // Now we have `S + S` and computed `S ^ F` locally.
+            let v = a_bit.clone() - &(a_bit.clone() * *b_bit * F::from(2));
+            let b_share = if b_bit.as_u128() == 1 {
+                one.clone()
+            } else {
+                S::ZERO
+            };
+            v + &b_share
+        })
```

Comment:
I don't understand why this is necessary. Isn't it just:

```
.map(|bit_index, a_bit| {
    if (b >> bit_index) & 1 == 0 {
        a_bit.clone()
    } else {
        one.clone() - a_bit.clone()
    }
}
```

Reply:
oh my god, so simple...